language/move-lang/src/parser/syntax.rs

```diff
@@ -103,25 +111,67 @@ fn consume_optional_token_with_loc<'input>(
     }
 }
 
+// Parse a comma-separated list of items, including the specified starting and
+// ending tokens.
 fn parse_comma_list<'input, F, R>(
     tokens: &mut Lexer<'input>,
-    list_end_token: Tok,
+    start_token: Tok,
+    end_token: Tok,
+    parse_list_item: F,
+    item_description: &str,
+) -> Result<Vec<R>, Error>
+where
+    F: Fn(&mut Lexer<'input>) -> Result<R, Error>,
+{
+    let start_loc = tokens.start_loc();
+    consume_token(tokens, start_token)?;
+    parse_comma_list_after_start(
+        tokens,
+        start_loc,
+        start_token,
+        end_token,
+        parse_list_item,
+        item_description,
+    )
+}
+
+// Parse a comma-separated list of items, including the specified ending token, but
+// assuming that the starting token has already been consumed.
+fn parse_comma_list_after_start<'input, F, R>(
+    tokens: &mut Lexer<'input>,
+    start_loc: usize,
+    start_token: Tok,
+    end_token: Tok,
     parse_list_item: F,
+    item_description: &str,
 ) -> Result<Vec<R>, Error>
 where
     F: Fn(&mut Lexer<'input>) -> Result<R, Error>,
 {
-    if tokens.peek() == list_end_token {
+    if match_token(tokens, end_token)? {
         return Ok(vec![]);
     }
     let mut v = vec![];
     loop {
+        if tokens.peek() == Tok::Comma || tokens.peek() == end_token {
+            let current_loc = tokens.start_loc();
+            let loc = make_loc(tokens.file_name(), current_loc, current_loc);
+            return Err(vec![(loc, format!("Expected {}", item_description))]);
+        }
         v.push(parse_list_item(tokens)?);
-        if tokens.peek() == list_end_token {
+        if match_token(tokens, end_token)? {
             break Ok(v);
         }
-        consume_token(tokens, Tok::Comma)?;
-        if tokens.peek() == list_end_token {
+        if !match_token(tokens, Tok::Comma)? {
+            let current_loc = tokens.start_loc();
+            let loc = make_loc(tokens.file_name(), current_loc, current_loc);
+            let loc2 = make_loc(tokens.file_name(), start_loc, start_loc);
+            return Err(vec![
+                (loc, format!("Expected '{}'", end_token.to_string())),
```

Comment:
You shouldn't need the `to_string` here if `Tok` implements `Display`.

If `Tok` doesn't, it should be easy to add